adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -24,20 +24,26 @@ import org.bdgenomics.adam.models.SequenceDictionary
 import org.bdgenomics.formats.avro.{ Contig, RecordGroupMetadata }
 import org.bdgenomics.utils.cli.SaveArgs
 
-trait GenomicRDD[T] {
+trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
 
   val rdd: RDD[T]
 
   val sequences: SequenceDictionary
+
+  def transform(tFn: RDD[T] => RDD[T]): U = {
+    replaceRdd(tFn(rdd))
+  }
+
+  protected def replaceRdd(newRdd: RDD[T]): U
 }
```

Comment:
This riffs off the spark.ml pipelines work: http://spark.apache.org/docs/latest/ml-guide.html#transformers

In:

```
val foos: FooRDD = sc.loadFoo("...")
val rdd: RDD[Foo] = foos.rdd
rdd.map(//... general Spark RDD methods
foos.adamSpecificMethods() // ADAM-specific methods from FooRDD
foos.saveAsBar()
```

You'd need to create a new `FooRDD` with the product of `rdd.map(...)`.